prober/derp.go

```diff
@@ -324,7 +442,20 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
 			}
 			switch v := m.(type) {
 			case derp.ReceivedPacket:
-				recvc <- v
+				if v.Source != fromc.SelfPublicKey() {
+					recvc <- fmt.Errorf("got data packet %d from unexpected source, %v", idx, v.Source)
+					return
+				}
+				if !bytes.Equal(v.Data, pkts[idx]) {
+					recvc <- fmt.Errorf("unexpected data packet %d: %q", idx, v.Data)
+					return
+				}
+				idx += 1
+				if idx == len(pkts) {
```

Comment:
I think this could be a bit less brittle if the success case closes the channel and we only send non-nil errors. That way the caller below can just spin on the channel til it closes, e.g.,

```go
defer close(recvc)
// ...
if idx == len(pkts) {
   return
}
// ...
var errs []error
for {
drain:
   select {
   case <-ctx.Done():
      return ...
   case err, ok := <-recvc:
      if !ok {
         break drain
      }
      errs = append(errs, err)
   }
}
return errors.Join(errs...)
```

or something like that (untested, but of the right general shape).

Whether we close or send nil doesn't matter so much, but if we expect multiple packets I think we have to drain the channel.

Actually if we _do_ close the channel when the goroutine exits, then we don't need to watch the ctx in the loop:

```go
for err := range recvc {
   errs = append(errs, err)
}
if ctx.Err() != nil {
   return ctx.Err()
}
return errors.Join(errs...)
```

would cover it.

Reply:
Thanks, used your version here.

Reply:
While testing this I needed to slightly re-work the way we run send & receive goroutines, and having to drain two channels with multiple errors concurrently was getting cumbersome. I went back to a `select` that breaks on the first error.